## 01. Importing Libraries

In [32]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02. Importing Data

In [33]:
# Turn project folder path into a string 
path = r'/Users/jerriejiang/Desktop/4.0/A4_Instacart Basket Analysis/02 Data/Original Data/'

In [34]:
path = r'/Users/jerriejiang/Desktop/4.0/A4_Instacart Basket Analysis'

In [35]:
# Import dataset orders_products_combined.pkl
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

In [36]:
# Import dataset products_checked.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'), index_col = False)

### 3. In a new notebook, import the orders_products_combined dataframe from the pickle file you just saved.

### 4. Check the shape of the imported dataframe (it should be the same as the one you exported—always check!).

In [37]:
# Check the output
df_ords_prods_combined.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           0   2539329        1    prior             1                   2   
2           0   2539329        1    prior             1                   2   
3           0   2539329        1    prior             1                   2   
4           0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered _merge  
0          0   both  
1          0   both  
2          0   both  
3          0   both  
4          0   both

In [38]:
df_ords_prods_combined.shape

(32434489, 12)

In [39]:
# Check the output
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [40]:
# Drop unnecessary columns
df_prods = df_prods.drop(['Unnamed: 0'], axis=1)

In [41]:
# Check the output
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [42]:
df_prods.shape

(49672, 5)

### 5. Determine a suitable way to combine the orders_products_combined dataframe with your products data set. Make sure you’re using your wrangled, cleaned, and deduped products data set stored in your “Prepared Data” folder from the previous Exercise’s task.

#### There's a mistake due to the same namde being used "merge". I will try another way to solve it. 

In [43]:
# Drop unnecessary columns
df_ords_prods_combined = df_ords_prods_combined.drop(['_merge'], axis=1)

In [44]:
# Check output
df_ords_prods_combined.head()

Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0           0   2539329        1    prior             1                   2   
1           0   2539329        1    prior             1                   2   
2           0   2539329        1    prior             1                   2   
3           0   2539329        1    prior             1                   2   
4           0   2539329        1    prior             1                   2   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  8                     NaN       14084                  2   
2                  8                     NaN       12427                  3   
3                  8                     NaN       26088                  4   
4                  8                     NaN       26405                  5   

   reordered  
0          0  
1          0  
2          0  
3          0  
4          0

In [45]:
df_ords_prods_combined.shape

(32434489, 11)

In [46]:
# Merge the updated dataframes
df_ords_prods_merge = df_prods.merge(df_ords_prods_combined, on = 'product_id', indicator = True)

In [47]:
# Check the output
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   Unnamed: 0  order_id  user_id eval_set  order_number  orders_day_of_week  \
0        1987   3139998      138    prior            28                   6   
1        1989   1977647      138    prior            30                   6   
2       11433    389851      709    prior             2                   0   
3       12198    652770      764    prior             1                   3   
4       12200   1813452      764    prior             3                   4   

   order_hour_of_day  days_since_prior_order  add_to_cart_order  reordered  \
0                 11                     3.0                  5          0   
1                 17                    20.0                  1          1   
2                 21                     6.0                 20          0   
3                 13                     NaN                 10          0   
4                 17                     9.0                 11          1   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [48]:
df_ords_prods_merge.shape

(32404859, 16)

### 6. Confirm the results of the merge using the merge flag.

In [49]:
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

#### The resulting data frame from the merge contains 32,404,859 rows, each with information from both input data sets, as we specifically chose an inner join for this project.

### 7. Export the newly created dataframe as ords_prods_merge in a suitable format (taking into consideration the size).

In [50]:
# Export data to pkl
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge.pkl'))